### DISCUSSION TITLE
Option to rename plausible.js script

### DISCUSSION THREAD

**Discussion by Empty2k12** [plausible/analytics#387] [Anything Else]
## Prerequisites
- [x] I have searched open and closed issues to make sure that the feature has not yet been requested.

## Feature request
Regex blocking is one common way, Ad/Script-Blockers use to filter out unwanted scripts. For example, domains and script names containing substrings like `stats`, `analytics` or `monitor` are blocked greedily by these algorithms. It's only a matter of time until the substring `plausible.js` starts appearing on those block-lists. Allowing CNAME cloaking (e.g. using custom domain with Plausible) is already an excellent step in preventing the script get blocked. However, the script-name could still be used for blocking. If there was a way to rename the script to `script.js`, `app.js`, `jquery.js` or something similar, detection rate could be lowered even further.

**How would you like it to work**
Either, the script is renamed to something more innocent for everybody (e.g. `app.js`), or we could allow configuring the script name in the site settings and the server would serve the script from the renamed source as well.

**Comment by ukutaht** [comment]
@Empty2k12 Thanks for the request.

For custom domains, you can already use any filename to load the javascript. `script.js`, `app.js` etc is valid if you're running a custom domain.

On the cloud domain (plausible.io) it's not possible but I don't think we're going to fool anyone by loading `plausible.io/js/script.js`. Blockers can just block the whole `plausible.io` domain fully.

**Reply by markcellus** [reply]
> Blockers can just block the whole `plausible.io` domain fully

AFAICT, it's already happening. I'm using Blockada and everything on plausible.io domain is blocked.

**Reply by metmarkosaric** [reply]
yeah, some adblockers block our domain name so you cannot even visit our website or read our blog posts. in any case, if you're worried about adblockers, we recommend you use the proxy https://plausible.io/docs/proxy/introduction

**Comment by Empty2k12** [comment]
> For custom domains, you can already use any filename to load the javascript. script.js, app.js etc is valid if you're running a custom domain.

That's good to know. Is it documented anywhere?

**Comment by Empty2k12** [comment]
I just tried loading as jquery.js, but I am getting the 404 page. Also, when trying app.js, the actual frontend code is served. What Am I doing wrong?

**Comment by ukutaht** [comment]
Where are you trying to load it from? Please provide the full URL. What I meant is that when you use a custom domain like `stats.yourdomain.com` then any JS filename works: `stats.yourdomain.com/js/app.js`

**Comment by Empty2k12** [comment]
That's what I am talking about.
We're using a self-hosted Plausible. It seems in that case the feature is not enabled, even though it's technically already a custom domain. I will try setting another subdomain from the Plausible GUI and see if that works.

**Comment by ukutaht** [comment]
Ah sorry I wasn't aware this is a self-hosted installation. Supporting this does require a code change. The server only supports loading `plausible.js` at the moment.

For custom domains on the Cloud version this is solved with a reverse proxy in front of the server but I don't think that's very relevant for self-hosted installations.

**Comment by oliver3** [comment]
Triggered by the reaction of @ukutaht, I've added the following rule to my reverse proxy nginx configuration
```
rewrite ^/js/pls.js$ /js/plausible.js last;
```
Not sure if this is the best way to do it, but I can now access the script with `/js/pls.js`. Maybe this can be added as a comment in https://github.com/plausible/hosting/blob/master/reverse-proxy/nginx/plausible ?

**Reply by FarisZR** [reply]
apache2 rules:

```
    RewriteRule ^/js/pls.js$ /js/plausible.js [PT]
    RewriteRule ^/js/pls2.js$ /js/plausible.hash.js [PT]
    RewriteRule ^/js/pls3.js$ /js/plausible.outbound-links.js [PT]
    RewriteRule ^/js/pls4.js$ /js/plausible.hash.outbound-links.js [PT]
```

**Reply by Dabako** [reply]
When I try the apache2 rules and edit my site to use /js/app.js (I used app.js instead of pls.js) It wont show me any analytics. 

```
        ProxyRequests OFF
        RewriteRule "^/js/app.js$" "/js/plausible.js" [PT]
        <Location />
                ProxyPreserveHost On
                ProxyPass http://localhost:8000/
                ProxyPassReverse http://localhost:8000/
        </Location>

        <Proxy *>
                Order deny,allow
                Allow from all
        </Proxy>
```

It seems that my website is requesting the app.js but it does not make the request back to the self hosted plausible server.

**Comment by hyperknot** [comment]
Based on @oliver3's answer, here is how the nginx part can look with the location block:

```
    rewrite ^/js/pl.js$ /js/plausible.js last;

    location / {
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_pass http://127.0.0.1:8000;
    }
```

Then you need to manually use `pl.js` instead of `plausible.js` in the HTML code.

**Comment by maxemann96** [comment]
Dropping also a traefik 2.4 configuration (with docker executor), terraform hashicorp language syntax, just normal docker container labels:

```hcl
labels {
  label = "traefik.http.routers.foo-router.middlewares"
  value = "plausible@docker"
}

labels {
  label = "traefik.http.middlewares.plausible.replacepathregex.regex"
  value = "^/js/IM_NOT_PLAUSIBLE\\.js"
}
labels {
  label = "traefik.http.middlewares.plausible.replacepathregex.replacement"
  value = "/js/plausible.js"
}
```

More information how this works [here](https://doc.traefik.io/traefik/middlewares/overview/) and [here](https://doc.traefik.io/traefik/middlewares/replacepathregex/)

**Reply by ukutaht** [reply]
Thanks everyone for crowd-sourcing solutions!

`plausible.js` was added to EasyList recently which means it's blocked by most adblockers. Adblocker people themselves have told me that their intention is not to block 1st party analytics like Plausible self-hosters. I think it makes sense in the next release to introduce a config option `TRACKER_NAME` and make it something generic by default.

**Comment by FarisZR** [comment]
this doesn't seem to really work
changing the js file name is not enough,
because adblockers will just block the /api/event call, which in turn stops visits from being counted.

tested with brave and ublock origin

**Reply by Miguelklappes** [reply]
Are you self hosted or in cloud?

**Reply by FarisZR** [reply]
self hosted

**Reply by Miguelklappes** [reply]
Then it should not be blocked, you are using your own domain, not the one from Plausible. Therefore no api call will be intercepted.

**Reply by FarisZR** [reply]
it doesn't work like that
adblockers block plausible.js regardless of the domain
same thing with /api/event call

**Reply by Miguelklappes** [reply]
Of course, but not if you change the name, I thought that was obvious in this context. I tested with all known adblockers, extension or not, and changing the name from _"plausible.js"_ to something like _"scriptwo.js"_ works.

**Reply by FarisZR** [reply]
i changed it to plus.js, and as i said, brave and ublock origin blocked the /api/event call

**Reply by hyperknot** [reply]
I can confirm that `/api/event` is **not** blocked currently when renaming the js file. I tested with default uBlock origin with EasyPrivacy.

**Reply by hyperknot** [reply]
If you use uBlock Origin, which filter list was blocking it?

**Reply by FarisZR** [reply]
![image](https://user-images.githubusercontent.com/35614734/128389880-9f1db936-8d0a-4cbe-b484-8a52e2674b60.png)
in brave

**Reply by FarisZR** [reply]
tried with a different domain, same result. 
but im going to try it again soon

**Reply by hyperknot** [reply]
uBlock Origin has the log window which tells you which filter blocked it.

**Reply by FarisZR** [reply]
well it turns out to be from the domain.
thanks for the ublock origin tip @hyperknot 
easy privacy has `/stats.*/event` rule

**Reply by hyperknot** [reply]
Great to know, it's an easy fix then!

**Reply by bradkane** [reply]
We were seeing the same issue with self-hosted.  In our case, analytics.domain.com was being blocked.  Going generic like server1.domain.com took care of it.

**Comment by itslevir** [comment]
I feel like hiding the name to app.js is very underhanded. It's like tricking people or users into thinking something it's not. Bad for plausible's and your reputation.

**Comment by gfra54** [comment]
Hello all,
Any news on this front ? Is the TRACKER_NAME  idea still a thing ? 
Would it help about adblocking software ?
I am on a self hosted instance, with a custom domain.
Thanks !
